server/src/test/java/org/elasticsearch/action/support/ListenerTimeoutsTests.java

```diff
@@ -0,0 +1,117 @@
+
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        Settings settings = Settings.builder().put(NODE_NAME_SETTING.getKey(), "node").build();
+        taskQueue = new DeterministicTaskQueue(settings, random());
+    }
+
+    public void testListenerTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        wrapped.onResponse(null);
+        wrapped.onFailure(new IOException("incorrect exception"));
+
+        assertFalse(completed.get());
+        assertThat(exception.get(), instanceOf(ElasticsearchTimeoutException.class));
+    }
+
+    public void testFinishNormallyBeforeTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
+        wrapped.onResponse(null);
+        wrapped.onFailure(new IOException("boom"));
+        wrapped.onResponse(null);
+
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        assertTrue(completed.get());
+        assertNull(exception.get());
+    }
+
+    public void testFinishExceptionallyBeforeTimeout() {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        AtomicReference<Exception> exception = new AtomicReference<>();
+        ActionListener<Void> listener = wrap(completed, exception);
+
+        ActionListener<Void> wrapped = ListenerTimeouts.wrapWithTimeout(taskQueue.getThreadPool(), listener, timeout, generic, "test");
+        wrapped.onFailure(new IOException("boom"));
+
+        assertTrue(taskQueue.hasDeferredTasks());
+        taskQueue.advanceTime();
+        taskQueue.runAllRunnableTasks();
+
+        assertFalse(completed.get());
+        assertThat(exception.get(), instanceOf(IOException.class));
+    }
+
+    private ActionListener<Void> wrap(AtomicBoolean completed, AtomicReference<Exception> exception) {
+        return new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void aVoid) {
+                assert completed.get() == false : "Should not be called twice";
+                completed.set(true);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                assert exception.get() == null : "Should not be called twice";
+                exception.set(e);
```

Comment:
same here, use `assertTrue(compareAndSet(false, true))` to also check that only one of onResponse / onFailure is called